### PR TITLE
[Safety] Replacing pointers with references/optional_ptr in the Binder

### PR DESCRIPTION
**Pull request by Tishj**
We were using a pointer to pass around `unique_ptr<ParsedExpression> &`'s that could potentially be updated, now we use a `reference` instead.

### CODE REVIEW

File: src/include/duckdb/optimizer/cse_optimizer.hpp [hunk 1]
```diff
@@ -30,7 +30,7 @@ class CommonSubExpressionOptimizer : public LogicalOperatorVisitor {
 	//! First iteration: count how many times each expression occurs
 	void CountExpressions(Expression &expr, CSEReplacementState &state);
 	//! Second iteration: perform the actual replacement of the duplicate expressions with common subexpressions nodes
-	void PerformCSEReplacement(unique_ptr<Expression> *expr, CSEReplacementState &state);
+	void PerformCSEReplacement(reference<unique_ptr<Expression>> expr, CSEReplacementState &state);
```
Comment (Mytherin) [inline src/include/duckdb/optimizer/cse_optimizer.hpp:33]: We should be able to use `&` instead here, no?

**Review by Mytherin — COMMENTED** [review]
Thanks! This has a few conflicts with #7131 but not everything.

There is no reason to use the `reference` class unless we need to re-assign the reference object. Can we use the standard reference `&` here instead?

File: src/include/duckdb/optimizer/cse_optimizer.hpp (hunk 1 above)
Comment (Tishj) [reply src/include/duckdb/optimizer/cse_optimizer.hpp:33]: Hmm, actually I think you're right, it was just a raw pointer before, we couldn't reassign it anyways.
I'll change it to a reference 👍

**Comment by Tishj** [comment]
> Thanks! This has a few conflicts with #7131 but not everything.
> 
> There is no reason to use the `reference` class unless we need to re-assign the reference object. Can we use the standard reference `&` here instead?

merged with that branch 👍

**Comment by Mytherin** [comment]
Can you merge with master? That should now include the branch

**Review by Mytherin — APPROVED** [review]
Thanks! All changes LGTM

**Comment by Mytherin** [comment]
Could you merge with master to restart the CI?

**Pull request merged**